src/app/home/home.component.ts

```diff
@@ -23,64 +34,104 @@ import { selectAllCountries } from '../store/country.selector';
     private router: Router,
-    private dialog: MatDialog
+    private dialog: MatDialog,
+    private route: ActivatedRoute
   ) {}
 
   ngOnInit(): void {
-    // Load countries from store dispatch
+    // Load countries from the store
     this.store.dispatch(loadCountries());
-    // this.countries$ = this.store.select(selectAllCountries);
 
-    //  filtered countries
-    this.filteredCountries$ = this.countries$;
+    // Handle filtering logic
+    this.filteredCountries$ = this.store.select(
+      selectFilteredCountries(this.searchTerm, this.searchBy)
+    );
 
-    // Handle search
-    this.filteredCountries$ = this.countryService.searchCountries(
-      this.searchTerm
-      );
+    this.routeSub = this.route.paramMap
+      .pipe(
+        switchMap((params: Params) => {
+          const countryNameWithHyphen = params['get']('name');
+          if (countryNameWithHyphen) {
+            const countryName = countryNameWithHyphen.replace(/-/g, ' ');
+            return this.store.select(selectCountryByName(countryName));
+          }
+          return [];
+        })
+      )
+      .subscribe((country) => {
+        if (country) {
+          this.openDialog(country);
+        }
+      });
   }
-
-  // open details dialog
-  openCountryDetail(country: Country): void {
-    // Replace spaces with hyphens in the country name
-    const countryNameWithHyphen = country.name.common.replace(/ /g, '-');
-
-    this.store.dispatch(selectCountry({ countryName: country.name.common }));
-
-    // Navigate to the route
-    this.router.navigate(['/country', countryNameWithHyphen]).then(() => {
-      const dialogRef = this.dialog.open(CountryDetailComponent, {
+  openDialog(country: Country): void {
+    if (!this.dialogRef) {
+      this.dialogRef = this.dialog.open(CountryDetailComponent, {
         width: '400px',
         data: { country },
       });
 
-      // After the dialog is closed, navigate to /
-      dialogRef.afterClosed().subscribe(() => {
-        this.router.navigate([''], {
-          queryParams: {},
-          queryParamsHandling: 'merge', // Merge to prevent overriding existing params
-        });
+      // Handle dialog close event
+      this.dialogRef.afterClosed().subscribe(() => {
+        this.dialogRef = null;
+        this.router.navigate(['']);
       });
-    });
+    }
+  }
+
+  // Open country detail dialog when clicking on a card
+  openCountryDetailDialog(country: Country): void {
+    const countryNameWithHyphen = country.name.common.replace(/ /g, '-');
+    this.store.dispatch(selectCountry({ countryName: country.name.common }));
+    this.router.navigate(['/country', countryNameWithHyphen]);
   }
-  //  search function
+
+  // Trigger search
   onSearch(): void {
-    // Filter countries based on search term
-    this.filteredCountries$ = this.countryService.searchCountries(
-      this.searchTerm
-    );
+    if (this.searchBy === 'name') {
+      this.store.dispatch(
```

Comment:
https://medium.com/@kavisha.talsania/rxjs-debouncetime-and-distinctuntilchanged-c9f11db4d45f


Please use the above approach for searching so that least amount of api calls occur